[ci] Add root pyproject.toml so isort behaves identically from any working directory

blauwers opened the pull request:
- fixes: isort non-deterministic behavior across repo root vs connector subdirectories
- Fixes #4798

### Proposed changes

* Add a project-wide `pyproject.toml` that pins `isort` configuration.
* Configure `isort` to consistently treat `.` as the source root (`src_paths = [.]`), ensuring identical results whether `isort` is run from the repository root or within a connector subdirectory.
* Align settings with OpenCTI's existing code style guidelines: `profile = black` matches the existing CI invocation (`isort --profile black --check .`) and the pre-commit hook (`isort --profile black --line-length 88`); `line_length = 88` is implied by the `black` profile and therefore not set explicitly.

### Verification

- `isort --profile black --check .` from the repository root produces zero diffs against current master.
- `isort --profile black --check src/main.py` from `external-import/pgl-yoyo/` (the example flagged in #4798) also produces zero diffs once the new config is in place. Without the config, the same command produces the spurious `from pycti import OpenCTIConnectorHelper` reorder diff documented in the issue.

### Related issues

* #4798

### Checklist

- [x] I consider the submitted work as finished
- [x] I have signed my commits using GPG key.
- [x] I tested the code for its functionality using different use cases
- [x] I added/update the relevant documentation (either on github or on notion)
- [x] Where necessary I refactored code to improve the overall quality

### Further comments

This change ensures that developers, CI, and pre-commit hooks all get deterministic `isort` behavior regardless of where in the repo the command is invoked. Without this, contributors see spurious diffs depending on their working directory, which complicates code reviews and wastes CI cycles.

SamuelHassine committed:
Add isort configuration to pyproject.toml (#4798)

``isort`` resolves first-party imports relative to its current working
directory when no project-wide configuration is found. As a result the
same file ends up with two different orderings depending on whether the
command is invoked from the repository root or from a connector
subdirectory — contributors see spurious diffs and CI cycles get wasted
re-running the formatter (see OpenCTI-Platform/connectors#4798).

Pin ``src_paths = [.]`` in a new root-level ``pyproject.toml`` so the
import resolution is identical regardless of the cwd. ``profile = black``
matches the existing CI invocation (``isort --profile black --check .``)
and the pre-commit hook (``isort --profile black --line-length 88``), and
``line_length = 88`` is implied by the ``black`` profile so it does not
need to be set explicitly.

Tested by running ``isort --profile black --check .`` from the repository
root and ``isort --profile black --check src/main.py`` from
``external-import/pgl-yoyo/`` — both produce zero diffs against the
current master, confirming the configuration is in line with the
existing imports.

Closes #4798.

SamuelHassine reviewed (dismissed):
LGTM — clean tooling fix. Reproduced the original ``isort`` non-determinism on master (running ``isort --profile black --check src/main.py`` from ``external-import/pgl-yoyo/`` flagged the documented ``from pycti import OpenCTIConnectorHelper`` reorder, while the same command from the repo root reported the file as clean) and confirmed that adding the root ``pyproject.toml`` from this PR makes both invocations agree, with zero diffs across the whole repo.

Tidied the commit message and PR body to spell out the rationale + verification, kept the configuration minimal (``profile = black``, ``src_paths = [.]``) so it stays aligned with the CI command ``isort --profile black --check .`` and the pre-commit hook (``isort --profile black --line-length 88``), and rebased onto current ``master`` so the squash applies cleanly.

CI is green (``base_linter``, ``ensure_formatting``, ``linter``, ``test``, ``build_manifest``, CLA, signed commits, PR title check) and ``mergeable: MERGEABLE``. Approving.

SamuelHassine committed:
[ci] Fix punctuation in pyproject.toml comment

Rewrite the comment so the sentence describing the bug ends with a
period (and a separate sentence carries the reference to the bug
report) instead of trailing off into a parenthetical without
punctuation. No semantic change to the `isort` configuration.

SamuelHassine commented on pyproject.toml:
```diff
@@ -0,0 +1,13 @@
+# connector subdirectory, or via the pre-commit hook. Without this file
+# ``isort`` resolves first-party imports relative to its current working
+# directory, so the same file ends up with a different ordering depending
+# on where the contributor (or CI) happens to run from
```
Fixed by 8cedfc0bba. The trailing sentence is rewritten so it ends with a proper period, and the bug-report reference moved to its own sentence:

```toml
# Repository-wide tooling configuration.
#
# Pinning ``src_paths`` here keeps ``isort`` output identical no matter
# whether the command is invoked from the repository root, from a
# connector subdirectory, or via the pre-commit hook. Without this file
# ``isort`` resolves first-party imports relative to its current working
# directory, so the same file ends up with a different ordering
# depending on where the contributor (or CI) happens to run from.
# See OpenCTI-Platform/connectors#4798 for the original bug report.

[tool.isort]
profile = "black"
src_paths = ["."]
```

No semantic change to the `isort` configuration itself; the comment now reads cleanly.

SamuelHassine reviewed (dismissed):
Re-approving after the Copilot nit fix, head commit 8cedfc0bba.

The single remaining Copilot thread is addressed and resolved:

- `pyproject.toml`: rewrite the trailing comment so the bug-description sentence ends with a period and the `#4798` reference moves to a separate sentence (`See OpenCTI-Platform/connectors#4798 for the original bug report.`). No semantic change to the `isort` configuration; the file still pins `profile = "black"` and `src_paths = ["."]`.

CI status on 8cedfc0bba:

- GitHub Actions: PR title convention, signed commits, do-not-merge, label / organisation check, issue-link — all green.
- CircleCI: `ensure_formatting`, `base_linter`, `linter`, `test` — all green. `build_manifest` is queued downstream of the four jobs above and consistently passed on the previous PR commit (`6b68084784` -> `build_num=467522` green); I do not expect it to surface anything new on a comment-only change.
- `filigran/cla` — already signed.

Verified locally that `isort --profile black --check .` from the repository root and `isort --profile black --check pyproject.toml internal-enrichment/anyrun-task/src/main.py` from `internal-enrichment/anyrun-task/` both produce zero diffs with the new config; the change is also a no-op for `black --check` (TOML is not formatted by black). Branch is `MERGEABLE` (no conflict with `master`).

SamuelHassine committed:
[ci] Document isort config precedence for connector-level overrides

Clarify the header comment in the root ``pyproject.toml`` so the
documented "consistent behaviour from any working directory" claim is
accurate even when three connectors ship their own
``[tool.isort]`` (``external-import/dragos``,
``external-import/proofpoint-tap``,
``external-import/tenable-security-center``).

``isort`` walks up the tree looking for the *closest* config and stops
at the first ``[tool.isort]`` it finds, so a connector-level section
takes precedence over this root file when ``isort`` is run from inside
that connector directory. The comment now spells this out and notes
that the existing CI invocation
(``isort --profile black --check .`` from the repository root) still
picks up this root config, which is where the consistency goal of
issue #4798 actually applies.

No behaviour change — comment-only update.

SamuelHassine commented on pyproject.toml:
```diff
@@ -0,0 +1,13 @@
+# Pinning ``src_paths`` here keeps ``isort`` output identical no matter
+# whether the command is invoked from the repository root, from a
+# connector subdirectory, or via the pre-commit hook. Without this file
+# ``isort`` resolves first-party imports relative to its current working
```
Good catch — fixed in 56a45675.

Went with the documentation-clarification path you suggested. The header comment in the root `pyproject.toml` now spells out that:

- `isort` walks up the tree looking for the *closest* `[tool.isort]` and stops there, so connector-level sections take precedence when the command is run from inside that connector;
- three connectors ship such an override on purpose (`external-import/dragos`, `external-import/proofpoint-tap`, `external-import/tenable-security-center`) because they bundle their own `black` / `ruff` / `mypy` settings and have a source layout that already produces the desired isort output;
- the consistency goal of issue #4798 is for the repo-wide CI invocation (`isort --profile black --check .` from the repository root), which still picks up the new root config — so the `pgl-yoyo` regression from the issue stays fixed.

I deliberately did **not** push `src_paths = ["."]` into the three connector-level sections in this PR. Adding it changes how those connectors classify their own `first-party` package (e.g. inside `external-import/dragos`, `from dragos.* import …` lines that today sort with the third-party block would move to a dedicated first-party block), which would re-sort ~16 unrelated files. That cleanup is better tackled in a follow-up scoped to each connector if/when their maintainers want to adopt the explicit-`src_paths` style.

SamuelHassine reviewed (approved):
Re-approving after the Copilot review fix, head commit `56a45675`.

The one remaining Copilot thread on the previous head (`8cedfc0bb`) is now addressed and resolved:

- `pyproject.toml`: header comment now documents the precedence rule explicitly. `isort` walks up the tree looking for the *closest* `[tool.isort]` and stops there, so the three connector-level overrides (`external-import/dragos`, `external-import/proofpoint-tap`, `external-import/tenable-security-center`) take precedence when `isort` is run from inside those connector directories. The comment also notes that the consistency goal of #4798 is for the repo-wide CI invocation (`isort --profile black --check .` from the repository root), which always picks up this root config — so the `pgl-yoyo` regression from the original bug report stays fixed.

I deliberately did **not** push `src_paths = ["."]` into the three connector-level sections. Adding it changes how those connectors classify their own first-party package (e.g. inside `external-import/dragos`, `from dragos.* import …` lines that today sort with the third-party block would move to a dedicated first-party block) and would re-sort ~16 unrelated files outside the scope of this PR. That cleanup is better tackled in a follow-up scoped to each connector if/when their maintainers want to adopt the explicit-`src_paths` style.

CI status on `56a45675`:

- GitHub Actions: PR title convention, signed commits, issue-link, do-not-merge — all green.
- CircleCI: `ensure_formatting`, `base_linter`, `linter` — all green.
- `filigran/cla` — signed.

Verification commands from the PR description both pass locally:

- `isort --profile black --check .` from the repository root → no diff against current master.
- `isort --profile black --check src/main.py` from `external-import/pgl-yoyo/` → no diff (the regression from #4798 stays fixed).
